csvy/validators.py

```diff
@@ -1,14 +1,91 @@
 """This module contains validators for the CSVY file format."""
 
 import csv
-from typing import Optional, TypeVar
+from typing import Any, Callable, Optional, TypeVar
 
 from pydantic import BaseModel, Field
 
+VALIDATORS_REGISTRY: dict[str, type[BaseModel]] = {}
+"""Registry of validators to run on the header."""
+
+
+def register_validator(
+    name: str, overwrite: bool = False
+) -> Callable[[type[BaseModel]], type[BaseModel]]:
+    """Registers a validator in the registry.
+
+    This function is a decorator that registers a validator in the registry. The name
+    of the validator is used as the key in the registry.
+
+    Args:
+        name: The name of the validator.
+        overwrite: Whether to overwrite the validator if it already exists.
+
+    Returns:
+        The decorator function that registers the validator.
+    """
+
+    def decorator(cls: type[BaseModel]) -> type[BaseModel]:
+        if not issubclass(cls, BaseModel):
+            raise TypeError("Validators must be subclasses of pydantic.BaseModel.")
+
+        if name in VALIDATORS_REGISTRY and not overwrite:
+            raise ValueError(f"Validator with name '{name}' already exists.")
+
+        VALIDATORS_REGISTRY[name] = cls
+        return cls
+
+    return decorator
```

Comment:
I'd usually use the `decorator` package for this to simplify the syntax a little: https://pypi.org/project/decorator/

But this is totally fine as is.

Reply:
I don't think this works well with classes, in particular in this case where we are just registering the class using an input to the decorator itself, not the function. So skipping, for now.